fiona/ogrext.pyx

```diff
@@ -255,18 +255,26 @@ cdef class FeatureBuilder:
         cdef void *cogr_geometry = NULL
 
         if not ignore_geometry:
-            cogr_geometry = OGR_F_GetGeometryRef(feature)
+            """ We steal the geometry: the geometry of the in-memory feature is now null
+            and we are responsible for cogr_geometry.
+            """
+            cogr_geometry = OGR_F_StealGeometry(feature)
 
             if cogr_geometry is not NULL:
-                code = OGR_G_GetGeometryType(cogr_geometry)
 
-                if 7 < code < 100:  # Curves.
+                code = base_geometry_type_code(OGR_G_GetGeometryType(cogr_geometry))
+
+                if 7 < code < 15:  # Curves.
                     cogr_geometry = get_linear_geometry(cogr_geometry)
```

Comment:
Now we have two geometries in memory because OGR_G_GetLinearGeometry makes a copy (the difference from the "ForceTo" functions is confusing!). For the curve situation, let's do something like

```
cdef OGRGeometryH original_geometry = cogr_geometry
cogr_geometry = get_linear_geometry(original_geometry)
OGR_G_DestroyGeometry(original_geometry)
```